pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,712 @@
+		/// amount_b = new_reserve_b - reserve_b
+		///
+		/// LP must have sufficient amount of asset a/b - amount_a and amount_b.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: liquidity provider
+		/// - `pool_id`: Pool Id
+		/// - `asset`: Asset id
+		/// - `amount`: liquidity amount of `asset` to be added to the pool.
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T::AssetId>,
+			asset: T::AssetId,
+			amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinTradingLimit::get(),
+				Error::<T>::InsufficientTradingAmount
+			);
+
+			// Ensure that who has enough balance of given asset
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+
+			ensure!(pool.contains_asset(asset), Error::<T>::AssetNotInPool);
+
+			// Load initial pool assets balances.
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+			let initial_reserves = AssetAmounts(
+				T::Currency::free_balance(pool.assets.0, &pool_account),
+				T::Currency::free_balance(pool.assets.1, &pool_account),
+			);
+
+			// Work out which asset is asset B ( second asset which has to provided by LP )
+			// Update initial reserves in correct order ( asset a is given as params, asset b is second asset)
+			let (assets, updated_a_reserve, initial_reserves) = if asset == pool.assets.0 {
+				let assets = (asset, pool.assets.1);
+				let reserves = AssetAmounts(initial_reserves.0, initial_reserves.1);
+				let updated_a_reserve = reserves.0.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+
+				(assets, updated_a_reserve, reserves)
+			} else {
+				let assets = (asset, pool.assets.0);
+				let reserves = AssetAmounts(initial_reserves.1, initial_reserves.0);
+				let updated_a_reserve = reserves.0.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+
+				(assets, updated_a_reserve, reserves)
+			};
+
+			// Calculate new reserve of asset b and work out the difference to get required amount of second asset which has to be provided by LP too.
+			let asset_b_reserve = calculate_asset_b_reserve(initial_reserves.0, initial_reserves.1, updated_a_reserve)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			let asset_b_amount = asset_b_reserve
+				.checked_sub(initial_reserves.1)
+				.ok_or(ArithmeticError::Underflow)?;
+
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= asset_b_amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let new_reserves = AssetAmounts(
+				updated_a_reserve,
+				initial_reserves
+					.1
+					.checked_add(asset_b_amount)
+					.ok_or(Error::<T>::InvalidAmplification)?,
```

Comment:
Why not use `asset_b_reserve`? What does it have to do with `InvalidAmplification`?

Reply:
Mistake and leftover from previous implementation.I'll correct.